merkletree.go

```diff
@@ -0,0 +1,270 @@
+package merkletree
+
+import (
+	"errors"
+
+	"github.com/coniks-sys/libmerkleprefixtree-go/crypto"
+	"github.com/coniks-sys/libmerkleprefixtree-go/internal"
+)
+
+var (
+	ErrInvalidTree = errors.New("[merkletree] invalid tree")
+)
+
+const (
+	EmptyBranchIdentifier = 'E'
+	LeafIdentifier        = 'L'
+)
+
+type MerkleTree struct {
+	treeNonce []byte
+	salt      []byte
+	privKey   []byte
+	pubKey    []byte
+	policies  Policies
+	root      *interiorNode
+}
+
+func InitMerkleTree(policies Policies, treeNonce, salt, pubKey, privKey []byte) *MerkleTree {
+	leftBranch := &emptyNode{
+		interiorNode: interiorNode{
+			level: 1,
+		},
+	}
+
+	rightBranch := &emptyNode{
+		interiorNode: interiorNode{
+			level: 1,
+		},
+	}
+	root := &interiorNode{
+		parent:     nil,
+		leftChild:  leftBranch,
+		rightChild: rightBranch,
+		leftHash:   nil,
+		rightHash:  nil,
+		level:      0,
+	}
+	leftBranch.parent = root
+	rightBranch.parent = root
+
+	m := &MerkleTree{
+		treeNonce: treeNonce,
+		salt:      salt,
+		privKey:   privKey,
+		pubKey:    pubKey,
+		policies:  policies,
+		root:      root,
+	}
+	return m
+}
+
+func LookUp(key string) (MerkleNode, []LookUpProofNode, error) {
+	str := getCurrentSTR()
+	return lookUp(key, str)
+}
+
+func LookUpInEpoch(key string, ep int64) (MerkleNode, []LookUpProofNode, error) {
+	str := GetSTR(ep)
+	return lookUp(key, str)
+}
+
+func lookUp(key string, str *SignedTreeRoot) (MerkleNode, []LookUpProofNode, error) {
```

Comment:
It seems like `lookUp` should be a method of a `MerkleTree`, no? The only thing you use from the `str` is the `treeRoot`.

Then, `LookUp` and `LookUpInEpoch` would be methods of the new `History` struct, which stores the cloned `MerkleTree`s.